[skip-ci][NFC][ntuple] add missing information to spec (#16704)
root-project/root: tree/ntuple/v7/doc/specifications.md

the field has been created as a virtual field from another, non-projected source field.
If a projected field has attached columns,
these columns are alias columns to physical columns attached to the source field.
The following restrictions apply on field projections:
  - The source field and the target field must have the same structure,

>     except for an `RNTupleCardinality` field that must have a collection field as a source.

Reviewer: ```suggestion
    except for an `RNTupleCardinality` field, which must have a collection field as a source.
```

>     except for an `RNTupleCardinality` field that must have a collection field as a source.

Reviewer 2: I'm not sure I fully understand the details here, but if I'm not mistaken for example the projected `RVec`s of the importer do not have the same subfield structure than the collections they point into. Is this covered here?

Author: clarified "structure" --> "structural role"

  - For streamer fields and leaf fields, the type name of the source field and the projected field must be identical.
  - Projections involving variants or fixed-size arrays are unsupported.

>   - Projected fields must be on the same schema path of collection fields than the source field.

Reviewer: ```suggestion
  - Projected fields must be on the same schema path of collection fields as the source field.
```

>     For instance, one can project a vector of struct with floats to individual vectors of floats but cannot

Reviewer: ```suggestion
    For instance, one can project a vector of structs with floats to individual vectors of floats but cannot
```
or alternatively, `vector of a struct`, but the suggestion is consistent with `vector of floats`.

## Glossary

### Anchor

> The anchor is a data block that makes the entry point to an RNTuple.

Reviewer 3: `makes` or `is`?

Author: changed `makes` to `represents`

> The anchor is specific to the RNTuple container in which the RNTuple data are embedded (e.g., a ROOT file).

Reviewer 3: ```suggestion
The anchor is specific to the RNTuple container in which the RNTuple data are embedded (e.g., a ROOT file or an object store).
```

The anchor must provide the information to load the header and the footer **envelopes**.

### Cluster

> A cluster is a set of **pages** that contain all the data belonging to a certain entry range.

Reviewer 3: ```suggestion
A cluster is a set of **pages** that contain all the data belonging to an entry range.
```

The data set is partitioned in clusters.
A typical cluster size is tens to hundreds of megabytes.

### Column

> A column is a storage backed vector of a number of **elements** of a simple type.

Reviewer 3: Probably good enough even-though it is semantically not quite true anymore (Streamed fields) (it is still technically true since it is backed by a bag of bytes ...)

Column elements have a fixed bit-length that depends on the column type.
Some column types allow setting the bit lengths within specific limits (e.g. for floats with truncated mantiassa).

### Envelope

An envelope is a data block with RNTuple meta-data, such as the header and the footer.

### Field

A field describes a serialized C++ type.
A field can have a hierarchy of subfields representing a composed C++ type (e.g., a vector of integers).

> A field has zero, one, or multiple **columns** attached to it.

Author: `std::string` and the streamer field.

> A field has zero, one, or multiple **columns** attached to it.

Reviewer 3: Isn't `std::string` stored the same as a collection of `char`? (if not, why not?)

> A field has zero, one, or multiple **columns** attached to it.

Author: It's not. Because `std::string` is an "atomic type" in contrast to a collection of char, which is a composed type.

## Glossary

### Anchor

The anchor is a data block that makes the entry point to an RNTuple.
The anchor is specific to the RNTuple container in which the RNTuple data are embedded (e.g., a ROOT file).
The anchor must provide the information to load the header and the footer **envelopes**.

### Cluster

> A cluster is a set of **pages** that contain all the data that belongs to a certain entry range.

Reviewer: ```suggestion
A cluster is a set of **pages** that contain all the data belonging to a certain entry range.
```

The data set is partitioned in clusters.
A typical cluster size is tens to hundreds of megabytes.

### Column

A column is a storage backed vector of a number of **elements** of a simple type.
Column elements have a fixed bit-length that depends on the column type.

> Some column types allow setting the bit lengths within limits, (e.g. for floats with truncated mantiassa).

Reviewer: ```suggestion
Some column types allow setting the bit lengths within specific limits (e.g. for floats with truncated mantiassa).
```

### Envelope

An envelope is a data block with RNTuple meta-data, such as the header and the footer.

### Field

A field describes a serialized C++ type.
A field can have a hierarchy of subfields representing a composed C++ type (e.g., a vector of integers).
A field has zero, one, or multiple **columns** attached to it.
The columns contain the data related to the field but not to its subfields, which have their own columns.

### Frame

> A frame is a byte range with meta-data information in an **envelope**.

Reviewer: ```suggestion
A frame is a byte range with metadata information in an **envelope**.
```
Other places in the text don't hyphenate. I don't have a specific preference, but we should be consistent.

## Notes on Backward and Forward Compatibility

Note that this section covers the backward and forward compatibility of the binary format itself.
It does not discuss schema evolution of the written types.

Readers supporting a certain version of the specification should support reading files
that were written according to previous versions of the same epoch.

> Readers should support reading data written according to _newer_ format versions of the same epoch in the following way

Reviewer: ```suggestion
Readers should support reading data written according to _newer_ format versions of the same epoch in the following way:
```

  - Unknown trailing information in the anchor, in envelopes, and in frames should be ignored.
    For instance, when reading frames, readers should continue reading after the frame-provided frame length
    rather than summing up the lengths of the known contents of the frame.

>     Checksum verification, however, should still take place and include known and unknown contents.

Reviewer: ```suggestion
    Checksum verification, however, should still take place and must include both known and unknown contents.
```

  - Unknown column, cluster, or field flags should be ignored.
  - Unknown IDs for extra type information should be ignored.

>   - When a reader encounters an unknown column type or an unknown field type or field version or field structure,

Reviewer: ```suggestion
  - When a reader encounters an unknown column type or an unknown field type, field version or field structure,
```

>     it should ignore the entire top-level field that the column or field belongs to.

Reviewer: ```suggestion
    it should ignore the entire top-level field this column or field belongs to.
```
